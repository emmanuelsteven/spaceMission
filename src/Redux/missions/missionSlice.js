import { createSlice, createAsyncThunk } from '@reduxjs/toolkit';
import axios from 'axios';
const initialState = {
    missions: [],
    isLoading: false,
    erro: null,
};

export const getMissions = createAsyncThunk ( 'missions/getMissions', async() => {
    try{
        const response = await axios.get('https://api.spacexdata.com/v3/missions');
        const allMissions = (response.data).map((mission) => mission);
        return allMissions;
    } catch (error){
        throw new Error ('failed to fetch mission data');
    }
});

const missionSlice = createSlice({
  name: 'missions',
  initialState,
  reducers: {
    joinMission: (state, action) => {
      state.missions.push(action.payload);
    
    },

    leaveMission: (state, action) => {
      state.missions = state.missions.filter((missions) => missions.id !== action.payload);
    },


    extraReducers: (builder) => {
        builder
        .addCase(getMissions.pending, (state) => {
            state.isLoading = true;
        })
        .addCase(getMissions.fulfilled, (state, {payload}) => {
            state.isLoading = false;

        })
        .addCase(getMissions.rejected, (state) => {
            state.isLoading = false;

        })
       
    }
  },
});

export const { joinMission, leaveMission } = missionSlice.actions;
export default missionSlice.reducer;
