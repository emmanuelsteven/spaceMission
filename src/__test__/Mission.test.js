// import React from 'react';
// import { render, fireEvent } from '@testing-library/react';
// import { Provider } from 'react-redux';
// import configureStore from 'redux-mock-store';
// import Mission from '../component/mission';
// import { onboardMission, offboardMission } from '../Redux/missions/missionSlice';

// const mockStore = configureStore([]);

// describe('Mission component', () => {
//   let store;
//   let missions;

//   beforeEach(() => {
//     missions = [
//       {
//         mission_id: 1,
//         mission_name: 'Mission 1',
//         description: 'Mission 1 Description',
//         reserved: false,
//       },
//       {
//         mission_id: 2,
//         mission_name: 'Mission 2',
//         description: 'Mission 2 Description',
//         reserved: true,
//       },
//     ];

//     store = mockStore({
//       mission: {
//         missions,
//         isLoading: false,
//       },
//     });
//   });

//   test('dispatches onboardMission when "Join Mission" button is clicked', () => {
//     store.dispatch = jest.fn();

//     const { getByText } = render(
//       <Provider store={store}>
//         <Mission />
//       </Provider>,
//     );

//     fireEvent.click(getByText('Join Mission'));
//     expect(store.dispatch).toHaveBeenCalledWith(onboardMission(1));
//   });

//   test('dispatches offboardMission when "Leave Mission" button is clicked', () => {
//     store.dispatch = jest.fn();

//     const { getByText } = render(
//       <Provider store={store}>
//         <Mission />
//       </Provider>,
//     );

//     fireEvent.click(getByText('Leave Mission'));
//     expect(store.dispatch).toHaveBeenCalledWith(offboardMission(2));
//   });
// });
import React from 'react';
import { render, fireEvent } from '@testing-library/react';
import { Provider } from 'react-redux';
import configureStore from 'redux-mock-store';
import renderer from 'react-test-renderer'; // Add this import
import Mission from '../component/mission';
import { onboardMission, offboardMission } from '../Redux/missions/missionSlice';

const mockStore = configureStore([]);

describe('Mission component', () => {
  let store;
  let missions;

  beforeEach(() => {
    missions = [
      {
        mission_id: 1,
        mission_name: 'Mission 1',
        description: 'Mission 1 Description',
        reserved: false,
      },
      {
        mission_id: 2,
        mission_name: 'Mission 2',
        description: 'Mission 2 Description',
        reserved: true,
      },
    ];

    store = mockStore({
      mission: {
        missions,
        isLoading: false,
      },
    });
  });

  test('dispatches onboardMission when "Join Mission" button is clicked', () => {
    store.dispatch = jest.fn();

    const { getByText } = render(
      <Provider store={store}>
        <Mission />
      </Provider>,
    );

    fireEvent.click(getByText('Join Mission'));
    expect(store.dispatch).toHaveBeenCalledWith(onboardMission(1));

    // Add snapshot testing
    const tree = renderer.create(
      <Provider store={store}>
        <Mission />
      </Provider>
    ).toJSON();
    expect(tree).toMatchSnapshot();
  });

  test('dispatches offboardMission when "Leave Mission" button is clicked', () => {
    store.dispatch = jest.fn();

    const { getByText } = render(
      <Provider store={store}>
        <Mission />
      </Provider>,
    );

    fireEvent.click(getByText('Leave Mission'));
    expect(store.dispatch).toHaveBeenCalledWith(offboardMission(2));

    // Add snapshot testing
    const tree = renderer.create(
      <Provider store={store}>
        <Mission />
      </Provider>
    ).toJSON();
    expect(tree).toMatchSnapshot();
  });
});
