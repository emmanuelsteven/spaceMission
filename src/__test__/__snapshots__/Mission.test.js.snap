// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Mission component dispatches offboardMission when "Leave Mission" button is clicked 1`] = `
<div
  className="mission_container"
>
  <table
    id="table"
  >
    <thead>
      <tr>
        <th>
          Mission
        </th>
        <th>
          Description
        </th>
        <th>
          Status
        </th>
        <th>
           
        </th>
      </tr>
    </thead>
    <tbody>
      <tr>
        <td
          className="mission_name"
        >
          Mission 1
        </td>
        <td
          className="mission_description"
        >
          Mission 1 Description
        </td>
        <td
          className="btns"
        >
          <button
            className="member-btn"
            type="button"
          >
            Not A Member
          </button>
        </td>
        <td
          className="btns"
        >
          <button
            className="join-btn"
            onClick={[Function]}
            type="button"
          >
            Join Mission
          </button>
        </td>
      </tr>
      <tr>
        <td
          className="mission_name"
        >
          Mission 2
        </td>
        <td
          className="mission_description"
        >
          Mission 2 Description
        </td>
        <td
          className="btns"
        >
          <button
            className="btnz"
            type="button"
          >
            Active Member
          </button>
        </td>
        <td
          className="btns"
        >
          <button
            className="leave-btn"
            onClick={[Function]}
            type="button"
          >
            Leave Mission
          </button>
        </td>
      </tr>
    </tbody>
  </table>
</div>
`;

exports[`Mission component dispatches onboardMission when "Join Mission" button is clicked 1`] = `
<div
  className="mission_container"
>
  <table
    id="table"
  >
    <thead>
      <tr>
        <th>
          Mission
        </th>
        <th>
          Description
        </th>
        <th>
          Status
        </th>
        <th>
           
        </th>
      </tr>
    </thead>
    <tbody>
      <tr>
        <td
          className="mission_name"
        >
          Mission 1
        </td>
        <td
          className="mission_description"
        >
          Mission 1 Description
        </td>
        <td
          className="btns"
        >
          <button
            className="member-btn"
            type="button"
          >
            Not A Member
          </button>
        </td>
        <td
          className="btns"
        >
          <button
            className="join-btn"
            onClick={[Function]}
            type="button"
          >
            Join Mission
          </button>
        </td>
      </tr>
      <tr>
        <td
          className="mission_name"
        >
          Mission 2
        </td>
        <td
          className="mission_description"
        >
          Mission 2 Description
        </td>
        <td
          className="btns"
        >
          <button
            className="btnz"
            type="button"
          >
            Active Member
          </button>
        </td>
        <td
          className="btns"
        >
          <button
            className="leave-btn"
            onClick={[Function]}
            type="button"
          >
            Leave Mission
          </button>
        </td>
      </tr>
    </tbody>
  </table>
</div>
`;
